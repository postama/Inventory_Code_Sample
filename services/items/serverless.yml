service: inventory-items

provider:
  name: aws
  runtime: nodejs8.10
  profile: kanso
  stage: local
  region: us-west-2
  environment:
    NODE_CONFIG_ENV: ${self:custom.stage}
  iamRoleStatements:
    - Effect: Allow
      Action:
        - dynamodb:Query
        - dynamodb:Scan
        - dynamodb:GetItem
        - dynamodb:PutItem
      Resource:
        - 'Fn::ImportValue': ${self:custom.stage}-InventoryTableArn
    - Effect: Allow
      Action:
        - dynamodb:Query
        - dynamodb:Scan
      Resource: {"Fn::Join": ["", [{"Fn::ImportValue":"${self:custom.stage}-InventoryTableArn"}, "/index/*"]]}

package:
  individually: ${self:custom.packageIndividually.${self:custom.stage}, self:custom.packageIndividually.other}

plugins:
  - serverless-add-api-key
  - serverless-webpack
  - serverless-offline
  - serverless-domain-manager

custom:
  apiKeys:
    - name: inventory-key
      value: ${ssm:/doorways/${self:custom.stage}/INVENTORY_API_KEY}
  customDomain:
    domainName: ${self:custom.domainPrefix.${self:custom.stage}, self:custom.domainPrefix.other}.doorways-services.net
    certificateName: '*.doorways-services.net'
    basePath: item
  domainPrefix:
    prod: inventory
    other: ${self:custom.stage}-inventory
  stage: ${opt:stage, self:provider.stage}
  region: ${opt:region, self:provider.region}
  packageIndividually:
    local: false
    other: true

functions:
  IssueItem:
    handler: issue-item/issue-item.entry
    events:
      - http:
          path: v1/{itemId}/command/issue
          method: POST
          private: true
  CreateItem:
    handler: create-item/create-item.entry
    events:
      - http:
          path: v1
          method: POST
          private: true
  EditItem: 
    handler: edit-item/edit-item.entry
    events:
      - http:
          path: v1/{itemId}
          method: PUT
          private: true
  DeleteItem:
    handler: delete-item/delete-item.entry
    events:
      - http:
          path: v1/{itemId}
          method: DELETE
          private: true
  AdjustItem:
    handler: adjust-qty/adjust-qty.entry
    events:
      - http:
          path: v1/{itemId}/command/adjust
          method: POST
          private: true
  ReceiveItem:
    handler: receive-item/receive-item.entry
    events:
      - http:
          path: v1/{itemId}/command/receive
          method: POST
          private: true
  DisposeItem:
    handler: dispose-item/dispose-item.entry
    events:
      - http:
          path: v1/{itemId}/command/dispose
          method: POST
          private: true
  TransferItem:
    handler: transfer-item/transfer-item.entry
    events:
      - http:
          path: v1/{itemId}/command/transfer
          method: POST
          private: true
