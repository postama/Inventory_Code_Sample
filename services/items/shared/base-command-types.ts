import { intersection, type, Int, partial } from 'io-ts';
import { UUID } from 'io-ts-types/lib/UUID'
import { BaseCommand } from '../../shared/base-command';
import { maxLength } from '../../../infrastructure/custom-brands/string-length';

export const BaseItemCommand = intersection([
    BaseCommand,
    type({ itemId: UUID })
]);

export const BaseQtyCommand = intersection([
    BaseItemCommand,
    type({ quantity: Int, locationId: UUID }),
    partial({ description: maxLength(200) })
]);
