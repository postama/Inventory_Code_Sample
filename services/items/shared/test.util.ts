import uuid from 'uuid/v4';
import { createBaseTestCommand } from '../../../infrastructure/test-util/test.util';
import { validateType } from '../../../infrastructure/validator/validator';

export function createTestInventoryCommand(type, validator, overrides = {}, base = {}) {
    return {
        ...validateType({
            ...base,
            ...createBaseTestCommand(type),
            itemId: uuid()
        }, validator), ...overrides
    };
}

export function createTestInventoryActionCommand(type, validator, overrides = {}, base = {}) {
    return {
        ...validateType({
            ...base,
            ...createBaseTestCommand(type),
            itemId: uuid(),
            quantity: 1,
            locationId: uuid(),
            description: 'test description'
        }, validator), ...overrides
    };
}