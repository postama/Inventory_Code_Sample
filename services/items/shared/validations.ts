import { getById } from '../../../repo/locations/location';
import { ValidationError } from '../../../../Kanso.Doorways.InventoryShared/errors';

export async function checkLocationExists(customerId: string, locationId: string, isActive = false): Promise<void> {
    const location = await getById(customerId, locationId);
    if (!location) throw new ValidationError(`Location ${locationId} does not exist`);
    if (isActive && !location.isActive) throw new ValidationError(`Location ${locationId} is not active`);
    return;
}