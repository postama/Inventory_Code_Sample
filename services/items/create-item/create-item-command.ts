import { type, TypeOf, literal, intersection } from 'io-ts';
import { validateType } from '../../../infrastructure/validator/validator';
import uuid from 'uuid/v4';
import { ItemDataValidator } from '../../../repo/items/item.type';
import { BaseItemCommand } from '../shared/base-command-types';

export const CreateItemCommandValidator = intersection([
    BaseItemCommand,
    type({ type: literal('CreateItem') }),
    ItemDataValidator
]);

export type CreateItemCommand = TypeOf<typeof CreateItemCommandValidator>;

export function createCreateItemCommand(request: any): CreateItemCommand {
    return validateType({ ...request, type: 'CreateItem', itemId: uuid() }, CreateItemCommandValidator);
}