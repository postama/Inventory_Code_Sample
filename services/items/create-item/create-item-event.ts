import { type, literal, TypeOf, intersection } from 'io-ts';
import { CreateItemCommand } from './create-item-command';
import { validateType } from '../../../infrastructure/validator/validator';
import { ItemDataValidator, createItemData } from '../../../repo/items/item.type';
import { BaseEventValidator, createBaseEvent } from '../../../repo/events/event-type';
import uuid from 'uuid/v4';

export const CREATE_ITEM = 'ItemCreated';

export const ItemCreatedValidatorV1 = intersection([
    BaseEventValidator,
    type({
        type: literal(CREATE_ITEM),
        version: literal(1),
        data: ItemDataValidator
    })
]);

export type ItemCreatedV1 = TypeOf<typeof ItemCreatedValidatorV1>;

export function createCreateItemEvent(command: CreateItemCommand): ItemCreatedV1 {
    const event = {
        ...createBaseEvent(command, uuid(), CREATE_ITEM, 1),
        data: createItemData(command)
    };

    return validateType(event, ItemCreatedValidatorV1);
}