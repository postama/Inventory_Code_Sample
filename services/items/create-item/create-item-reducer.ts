import { Item } from '../../../repo/items/item.type';
import { ItemCreatedV1 } from './create-item-event';

// _item is unused but still passed in. Starting with _ makes the compiler not complain about it being unused, but makes tslint complain about naming
// tslint:disable-next-line variable-name
export function createItemReducer(_item: Item, event: ItemCreatedV1): Item {
    return {
        ...event.data,
        quantity: 0,
        created: event.created,
        locations: {}
    };
}
