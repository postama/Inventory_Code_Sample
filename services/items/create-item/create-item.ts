import 'source-map-support/register';
import { save } from '../../../repo/items/item';
import { checkUniqueCommand } from '../../../repo/events/event';
import { ItemCreatedV1, createCreateItemEvent } from './create-item-event';
import { CreateItemCommand, createCreateItemCommand } from './create-item-command';
import { APIGatewayEvent } from 'aws-lambda';
import { createItemReducer } from './create-item-reducer';
import { requestHandler } from '../../../../Kanso.Doorways.InventoryShared/entry-wrapper/entry-wrapper';
import { checkLocationExists } from '../shared/validations';

export async function entry(request: APIGatewayEvent) {
    return requestHandler(createItem, request);
}

export async function createItem(request): Promise<{ id: string }> {
    const command: CreateItemCommand = createCreateItemCommand(request);
    await Promise.all([
        command.primaryLocation && checkLocationExists(command.customerId, command.primaryLocation),
        checkUniqueCommand(command)
    ]);
    const event: ItemCreatedV1 = createCreateItemEvent(command);
    event.state = createItemReducer({} as any, event);
    await save(event);
    return { id: event.entityId };
}