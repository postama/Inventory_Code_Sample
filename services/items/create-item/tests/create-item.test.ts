import uuid from 'uuid/v4';
import { entry, createItem } from '../create-item';
import { createItemReducer } from '../create-item-reducer';
import { createTestCreateItemEvent, createTestCreateItemCommand } from './test.util';
import { createCreateItemCommand } from '../create-item-command';
import { createCreateItemEvent } from '../create-item-event';
import { getById } from '../../../../repo/items/item';

import * as entryLib from '../../../../../Kanso.Doorways.InventoryShared/entry-wrapper/entry-wrapper';
jest.mock('../../../../../Kanso.Doorways.InventoryShared/entry-wrapper/entry-wrapper');
const entryMock = entryLib as jest.Mocked<typeof entryLib>;

import * as validator from '../../shared/validations';
jest.mock('../../shared/validations.ts');
const validatorMock = validator as jest.Mocked<typeof validator>;

import * as inventoryRepo from '../../../../repo/items/item';
const spy = jest.spyOn(inventoryRepo, 'save');

describe('create inventory item', () => {
    describe('command structure validator', () => {
        test('should fail with missing field with human readable error messages', () => {
            const command = createTestCreateItemCommand({ user: undefined });
            expect.assertions(2);
            try {
                createCreateItemCommand(command);
            } catch (e) {
                expect(e.name).toBe('ValidationError');
                expect(e.message).toBe('Expecting string at user but instead got: undefined.');
            }
        });

        test('should pass without optional fields', () => {
            const command = createTestCreateItemCommand({ supplier: undefined });
            const validCommand = createCreateItemCommand(command);
            expect(validCommand).toMatchSnapshot({
                itemId: expect.any(String),
                customerId: expect.any(String),
                commandId: expect.any(String),
                siteId: expect.any(String)
            });
        });

        test('should pass with full structure', () => {
            const command = createTestCreateItemCommand();
            const validCommand = createCreateItemCommand(command);
            expect(validCommand).toMatchSnapshot({
                itemId: expect.any(String),
                customerId: expect.any(String),
                commandId: expect.any(String),
                siteId: expect.any(String)
            });
        });
    });

    describe('event creator', () => {
        test('should throw if is given a bad command', () => {
            expect.assertions(2);
            try {
                const createItemCommand = createTestCreateItemCommand({ user: undefined });
                createCreateItemEvent(createItemCommand);
            } catch (e) {
                expect(e.name).toBe('ValidationError');
                expect(e.message).toBe('Expecting string at user but instead got: undefined.');
            }
        });

        test('should create an event missing optional fields', () => {
            const createItemCommand = createTestCreateItemCommand({ supplier: undefined });
            const event = createCreateItemEvent(createItemCommand);
            expect(event.type).toBe('ItemCreated');
        });

        test('should create an event if issued a valid command', () => {
            const createItemCommand = createTestCreateItemCommand();
            const event = createCreateItemEvent(createItemCommand);
            expect(event.type).toBe('ItemCreated');
        });

        test('unsupported fields should be stripped', () => {
            const createItemCommand = createTestCreateItemCommand({ toStrip: 'stripMe' });
            const event: any = createCreateItemEvent(createItemCommand);
            expect(event.toStrip).toBeUndefined();
        });
    });

    describe('event creator - integration', () => {
        test('should successfully create the item', async () => {
            const command = createTestCreateItemCommand();
            await createItem(command);
            const itemId = spy.mock.calls[0][0].entityId;
            const item = await getById(command.customerId, itemId);
            expect(item.quantity).toBe(0);
        });

        test('should pass when a primary location exists - integration', async () => {
            const command = createTestCreateItemCommand({ primaryLocation: uuid() });
            validatorMock.checkLocationExists.mockResolvedValue();
            await createItem(command);
            const itemId = spy.mock.calls[1][0].entityId;
            const item = await getById(command.customerId, itemId);
            expect(item.primaryLocation).toBe(command.primaryLocation);
            spy.mockReset();
        });
    });

    describe('reducer', () => {
        test('should create an inventory item from an event', () => {
            const event = createTestCreateItemEvent();
            const item = createItemReducer({} as any, event);
            expect(item).toMatchSnapshot({ created: expect.any(String) });
        });
    });

    describe('entry', () => {
        test('should call the entry handler', async () => {
            const command = createTestCreateItemCommand();
            await entry({ body: JSON.stringify(command) } as any);
            expect(entryMock.requestHandler).toBeCalled();
        });
    });
});
