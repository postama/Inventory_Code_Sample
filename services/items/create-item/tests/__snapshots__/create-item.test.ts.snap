// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create inventory item command structure validator should pass with full structure 1`] = `
Object {
  "brand": "abc",
  "commandId": Any<String>,
  "costPerUnit": 123,
  "customerId": Any<String>,
  "group": "abc",
  "identifier": "abc",
  "isPreferred": true,
  "itemId": Any<String>,
  "name": "abc",
  "siteId": Any<String>,
  "supplier": "abc",
  "type": "CreateItem",
  "unitOfMeasurement": "abc",
  "user": "abc",
}
`;

exports[`create inventory item command structure validator should pass without optional fields 1`] = `
Object {
  "brand": "abc",
  "commandId": Any<String>,
  "costPerUnit": 123,
  "customerId": Any<String>,
  "group": "abc",
  "identifier": "abc",
  "isPreferred": true,
  "itemId": Any<String>,
  "name": "abc",
  "siteId": Any<String>,
  "supplier": undefined,
  "type": "CreateItem",
  "unitOfMeasurement": "abc",
  "user": "abc",
}
`;

exports[`create inventory item reducer should create an inventory item from an event 1`] = `
Object {
  "brand": "abc",
  "costPerUnit": 123,
  "created": Any<String>,
  "group": "abc",
  "identifier": "abc",
  "isPreferred": true,
  "locations": Object {},
  "name": "abc",
  "quantity": 0,
  "supplier": "abc",
  "unitOfMeasurement": "abc",
}
`;
